src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,53 +1,103 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 /**
  * <p>Реалізувати свій ArrayList який імплементує інтерфейс List. Дотриматися основних вимог щодо
  * реалізації ArrayList (default capacity, newCapacity...)</p>
  */
 public class ArrayList<T> implements List<T> {
 
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] values;
+    private int size;
+
+    public ArrayList() {
+        values = (T[]) new Object[DEFAULT_CAPACITY];
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
-
+        extend();
+        values[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayIndexOutOfBoundsException();
+        }
+        extend();
+        System.arraycopy(values, index, values, index + 1, size - index);
+        values[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        isValid(index);
+        return values[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        isValid(index);
+        values[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        isValid(index);
+        T element = values[index];
+        System.arraycopy(values, index + 1, values, index, size - index);
```

Comment:
```suggestion
        System.arraycopy(values, index + 1, values, index, size - index - 1);
```
we should move one element less here because we are removing